pdr_backend/pdr_dashboard/util/format.py

```diff
@@ -111,74 +123,98 @@ def format_df(
     columns = df.columns
 
     if "sales" in columns and "df_buy_count" in columns and "ws_buy_count" in columns:
-        if df.is_empty():
-            df = df.with_columns(pl.col("sales_str").alias("sales_str"))
-        else:
-            df = df.with_columns(
-                sales_str=pl.struct(
-                    ["sales", "df_buy_count", "ws_buy_count"]
-                ).map_elements(
-                    lambda x: format_sales_info_data(
-                        x["sales"], x["df_buy_count"], x["ws_buy_count"]
-                    ),
-                    return_dtype=pl.String,
-                )
-            )
+        df = format_sales_info_data(df)
 
     for col in columns:
         if col == "sales_str":
             continue
 
-        df = df.with_columns(
-            # pylint: disable=cell-var-from-loop
-            pl.col(col)
-            .map_elements(lambda x: format_value(x, col), return_dtype=pl.String)
-            .alias(col)
-        )
+        df = format_column(df, col)
 
     return df
 
 
 @enforce_types
-def format_sales_info_data(total, df_buy, ws_buy) -> str:
-    total_str = format_currency(total, suffix="", show_decimal=False)
+def format_eth_address(df: pl.DataFrame, col: str) -> pl.DataFrame:
+    return df.with_columns(
+        pl.when(pl.col(col).str.len_chars() > 0)
+        .then(pl.col(col).str.slice(0, 5) + "..." + pl.col(col).str.slice(-5))
+        .otherwise(pl.lit("No address"))
+        .alias(col)
+    )
 
-    df_buy_str = (
-        f"{format_currency(df_buy, suffix='', show_decimal=False)} DF"
-        if df_buy > 0
-        else ""
+
+@enforce_types
+def format_percentage(df: pl.DataFrame, col: str) -> pl.DataFrame:
+    return df.with_columns(
+        (pl.col(col).round(2).cast(pl.String) + pl.lit("%")).alias(col)
     )
-    ws_buy_str = (
-        f"{format_currency(ws_buy, suffix='', show_decimal=False)} WS"
-        if ws_buy > 0
-        else ""
+
+
+@enforce_types
+def format_sales_info_data(df: pl.DataFrame) -> pl.DataFrame:
```

Comment:
It could be I missed what you do, I am not sure but it might be like the following:

```
def format_sales_info_data(df: pl.DataFrame) -> pl.DataFrame:
    if df.is_empty():
        return df.with_columns(pl.col("sales_str").alias("sales_str"))

    total = format_currency(df, "sales")["sales"]
    df_buy = format_currency(df, "df_buy_count", suffix=" DF")["df_buy_count"]
    ws_buy = format_currency(df, "ws_buy_count", suffix=" WS")["ws_buy_count"]

    # Combine all operations and directly assign to sales_str
    df = df.with_columns(
        [
            pl.when(pl.col("df_buy_count") > 0).then(df_buy).otherwise(pl.lit("")).alias("df_buy_str"),
            pl.when(pl.col("ws_buy_count") > 0).then(ws_buy).otherwise(pl.lit("")).alias("ws_buy_str"),
            
            # Create sales_str
            (
                total
                + pl.lit(" (")
                + pl.concat_str([pl.col("df_buy_str"), pl.col("ws_buy_str")], separator=", ")
                + pl.lit(")")
            )
            .str.replace("( )", "")  # If both df_buy_str and ws_buy_str are empty, remove the parentheses
            .alias("sales_str")
        ]
    )

    return df
```

Reply:
it's another two-step, unfortunately. When using pl.col("df_buy_str") it's the one defined in the same operation.